src/algorithms/fardetectors/FarDetectorLinearTracking.cc

```diff
@@ -74,42 +80,65 @@ void FarDetectorLinearTracking::process(const FarDetectorLinearTracking::Input&
       trace("No hits in layer");
       return;
     }
-    convertedHits.push_back(ConvertClusters(*layerHits));
+    ConvertClusters(*layerHits, *assocHits, convertedHits, assocParts);
   }
 
   // Create a matrix to store the hit positions
   Eigen::MatrixXd hitMatrix(3, m_cfg.n_layer);
-  // Loop over all combinations of hits fitting a track to all layers
-  buildMatrixRecursive(m_cfg.n_layer - 1, &hitMatrix, convertedHits, outputTracks);
-}
 
-void FarDetectorLinearTracking::buildMatrixRecursive(
-    int level, Eigen::MatrixXd* hitMatrix, const std::vector<std::vector<Eigen::Vector3d>>& hits,
-    gsl::not_null<edm4eic::TrackSegmentCollection*> outputTracks) const {
+  // Create vector to store indexes of hits in the track
+  std::vector<int> layerHitIndex(m_cfg.n_layer, 0);
 
-  // Iterate over hits in this layer
-  for (auto hit : hits[level]) {
+  int layer = 0;
 
-    hitMatrix->col(level) << hit;
+  // Iterate over all combinations of measurements in the layers without recursion
+  while (layer >= 0) {
```

Comment:
```suggestion
  while (true) {
```
Might as well document that it's up to `break` to terminate this loop.